ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -143,29 +146,72 @@ protected void applyUndelete(MessageInfo remoteBlob, RemoteReplicaInfo remoteRep
    */
   @Override
   protected void afterHandleReplicaMetadataResponse(RemoteReplicaGroup group) {
-    List<ExchangeMetadataResponse> exchangeMetadataResponseList = group.getExchangeMetadataResponseList();
-    List<RemoteReplicaInfo> replicasToReplicatePerNode = group.getRemoteReplicaInfos();
+    group.setState(ReplicaGroupReplicationState.DONE);
+  }
+
+  protected MessageInfo mapBlob(MessageInfo blob) {
+    StoreKey keyConvert = null;
     try {
-      EnumSet<MessageInfoType> acceptableLocalBlobStates = EnumSet.of(MessageInfoType.PUT);
-      EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
-      for (int i = 0; i < exchangeMetadataResponseList.size(); i++) {
-        ExchangeMetadataResponse exchangeMetadataResponse = exchangeMetadataResponseList.get(i);
-        RemoteReplicaInfo remoteReplicaInfo = replicasToReplicatePerNode.get(i);
-        if (exchangeMetadataResponse.serverErrorCode == ServerErrorCode.No_Error) {
-          for (MessageInfo messageInfo : exchangeMetadataResponse.getMissingStoreMessages()) {
-            // Check local store once before logging an error
-            checkLocalStore(messageInfo, remoteReplicaInfo, acceptableLocalBlobStates, acceptableStoreErrorCodes);
-          }
-          // Advance token so that we make progress in spite of missing keys,
-          // else the replication code will be stuck waiting for missing keys to appear in local store.
-          advanceToken(remoteReplicaInfo, exchangeMetadataResponse);
-        }
-      }
-    } catch (Exception e) {
-      logger.error("Fail to check Local store", e);
-    } finally {
-      group.setState(ReplicaGroupReplicationState.DONE);
+      // Don't do batch-convert, if one replica in batch fails, then it affects handling others
+      keyConvert = storeKeyConverter.convert(Collections.singleton(blob.getStoreKey()))
+          .get(blob.getStoreKey());
+    } catch (Throwable e) {
+      metrics.backupIntegrityError.inc();
+      logger.error("Failed to convert blobID {} due to ", blob.getStoreKey().getID(), e.toString());
     }
+    return new MessageInfo(keyConvert, blob);
+  }
+
+  /**
+   * Checks each blob from server with its counterpart in cloud backup.
+   * There are 4 cases.
+   * server-blob : cloud-blob
+   * absent : absent  => nothing to do
+   * absent : present => print to file, blob may be compacted or missing on server
+   * present: absent  => print to file, blob may not have been backed up or somehow missing in cloud backup
+   * present: present => check blob-state matches, if not then print to file
+   * @param response The {@link ReplicaMetadataResponse}.
+   * @param replicas
+   * @param server The remote {@link DataNodeId}.
+   * @return
+   */
+  List<ExchangeMetadataResponse> handleReplicaMetadataResponse(ReplicaMetadataResponse response,
+      List<RemoteReplicaInfo> replicas, DataNodeId server) {
+    IntStream.range(0, response.getReplicaMetadataResponseInfoList().size())
```

Comment:
A for loop would be better than a IntStream.range.

Reply:
So we need the index i because we are iterating two lists - response-list and replica-list. So the for loop would iterate using indices like for (i=0; i < size; ++i) and IntStream is the same and a little cleaner, don't you think ?

Reply:
done